Add ZStack

projhect opened the pull request:
Implemented `ZStack` mostly copying the code from `HStack` and `VStack`.

<details><summary><b>Example 1</b></summary>

```
  var body: some View {
    ZStack(alignment: Alignment(horizontalAlignment: .leading, verticalAlignment: .top)) {
        HStack {
            Spacer().frame(width: 1)
            Text( To-do ")
        }
      ToDoList()
        .padding()
        .border()
        .background(.blue)
    }
  }
}
```

![image](https://user-images.githubusercontent.com/121196946/209483888-a66b7d04-16a3-4671-bb69-ea2ff8b4efff.png)

</details>

<details><summary><b>Example 2</b></summary>

```
  let background = """
    

    |￣￣￣￣￣￣￣|
    |              |
    |＿＿＿＿＿_＿_|
    (\\__/) ||
    (•ㅅ•) ||
    / 　 づ
    """
  var body: some View {
    ZStack(alignment: .center) {
      Text("Broccoli ♡♡♡")
      VStack {
        ForEach(background.components(separatedBy: "\n"), id: \.self) { line in
          Text(line)
        }
      }
    }
  }
```

![image](https://user-images.githubusercontent.com/121196946/209483548-3a207572-eec2-4210-af85-3922e94f70f1.png)

</details>

projhect committed:
Add ZStack

rensbreur commented on Sources/SwiftTUI/View/Views/Stacks/ZStack.swift:
```diff
@@ -0,0 +1,116 @@
+
+    override func size(proposedSize: Size) -> Size {
+        var size: Size = .zero
+        for control in children.sorted(by: { $0.verticalFlexibility(width: proposedSize.width) < $1.verticalFlexibility(width: proposedSize.width) }) {
```
I think we can remove the use of `verticalFlexibility` here and in the `layout` method.

rensbreur commented on Sources/SwiftTUI/View/Views/Stacks/ZStack.swift:
```diff
@@ -0,0 +1,116 @@
+
+    func loadData(node: Node) {
+        for i in 0 ..< node.children[0].size {
+            (node.control as! ZStackControl).addSubview(node.children[0].control(at: i), at: i)
```
I noticed that the contents of the `ZStack` are drawn in the wrong order. I think that would be confusing for anyone trying to use it. Not sure what's the best way of dealing with this, we might be able to get away with changing the order in which any control draws its children, but that would require some testing if nothing else breaks.

rensbreur replied:
I think this will break `ZStack`s with dynamic content (e.g. with an `if` statement in it).

projhect replied:
That's true.

I validated the changes with what I thought was a moderately complex drop down menu. But in my example code the dynamic content was in the children of the `ZStack`.

I checked now and, as you said, adding an `if` directly to the `ZStack` content does not work.

projhect replied:
With the new set of changes, I observed dynamic content issues seem to be fixed.

rensbreur commented on Sources/SwiftTUI/View/Views/Stacks/ZStack.swift:
```diff
@@ -0,0 +1,116 @@
+    public let content: Content
+    let alignment: Alignment
+
+    public init(alignment: Alignment = .center, @ViewBuilder _ content: () -> Content) {
```
For `VStack` and `HStack` I've chosen a default alignment of top-leading (unlike SwiftUI) as I expected this would be more common and desired for text-based interfaces. I'd suggest we use that here as well, until a decision is made to change to a default center alignment for all views.

rensbreur reviewed (commented):
Hi @projhect, thanks for wanting to contribute! I like the look of the to-do panel you created using the ZStack. 

I would like to have the drawing order fixed before merging this in.

projhect committed:
Set ZStack default alignment

projhect committed:
Remove verticalFlexibility from ZStack size and layout

projhect committed:
Fix ZStack drawing order

projhect commented:
Thank you for reviewing the changes.

I have applied the fixes you suggested. So far, I didn't notice any issues with reversing the node children.

projhect committed:
Fix ZStacks with dynamic content

projhect committed:
Override horizontal selection methods in ZStackControl

projhect committed:
Make ZStack double-check firstResponder after removing subviews

projhect commented:
Your idea of changing the order in which any control draws its children is way beyond my current understanding of the library. I already bit off more than I can chew with the `ZStack` alone (without modifying the rest of the library) 😓 

Anyways, I wanted to give it one last try.

This time I attempted to keep all the changes related with the view order in `ZStackControl` (i.e. `addSubview` and `removeSubview` overrides).

I also added horizontal selection overrides by copying `HStack` ones (inverted). Perhaps `HStack` and `VStack` approaches to selection don’t suit that well `ZStack`, but it’s a start.

Finally, I added a check that resigns first responder, if needed, after calling `super.removeSubview(at:)`. That check probably does not belong there, though. Because the issue is not unique to ZStack and could be addressed in a more general way.

rensbreur commented:
Thanks for taking the time to make those changes! Based on your work, I've pushed a commit that makes use of the drawing order being reversed everywhere, I've added you as a co-author. I've also removed some more unused lines in the layout method. Checking for selection on `removeSubview`, should already be checked, so adding it to `ZStack` shouldn't be necessary.
